Have functions that only require VMs take a VM reference https://bugs.webkit.org/show_bug.cgi?id=202392

LumioseSil opened the pull request:
#### 7a682381645f03d8fc7f2fa58c134d45b5b59c2d
<pre>
Have functions that only require VMs take a VM reference <a href="https://bugs.webkit.org/show_bug.cgi?id=202392">https://bugs.webkit.org/show_bug.cgi?id=202392</a>

Reviewed by Darin Adler.

We can have the .vm() method be called before these functions,
and pass the result VM reference instead of passing the pointer to the JSGlobalObject, potentially calling the .vm() methods redundantly in each method to which said JSGlobalObject pointer is passed, and dereferencing the pointers to the VM.

* Source/JavaScriptCore/runtime/JSONObject.cpp:(value): Change argument
  type from JSGlobalObject* to VM&amp;.

* Source/WebCore/Modules/encryptedmedia/MediaKeyStatusMap.cpp:
  (get): Pass the VM of the state object to convertEnumerationToJS
  rather than the whole object.

* Source/WebCore/bindings/js/JSDOMConvertEnumeration.h:
  (convertEnumerationToJS): Change argument type from JSGlobalObject to
  VM.
  (convert): Pass the VM of lexicalGlobalObject instead of the entire
  global object.

* Source/WebCore/bindings/scripts/CodeGeneratorJS.pm: Ditto.

* Source/WebCore/bindings/scripts/test/JS/JSTestCallbackInterface.cpp:
  Ditto.

* Source/WebCore/bindings/scripts/test/JS/JSTestCallbackInterface.h:
  Ditto.

* Source/WebCore/bindings/scripts/test/JS/JSTestDefaultToJSONEnum.cpp:
  Ditto.

* Source/WebCore/bindings/scripts/test/JS/JSTestDefaultToJSONEnum.h:
  Ditto.

* Source/WebCore/bindings/scripts/test/JS/JSTestObj.cpp: Ditto.

* Source/WebCore/bindings/scripts/test/JS/JSTestObj.h: Ditto.

* Source/WebCore/bindings/scripts/test/JS/JSTestStandaloneDictionary.cpp:
  Ditto.

* Source/WebCore/bindings/scripts/test/JS/JSTestStandaloneDictionary.h:
  Ditto.

* Source/WebCore/bindings/scripts/test/JS/JSTestStandaloneEnumeration.cpp:
  Ditto.

* Source/WebCore/bindings/scripts/test/JS/JSTestStandaloneEnumeration.h:
  Ditto.

Canonical link: <a href="https://commits.webkit.org/262166@main">https://commits.webkit.org/262166@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9bfa4bb763403077e3966f26e50d9a9513b298e2

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/805 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/829 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/859 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/1093 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/712 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/794 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/877 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/912 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/932 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/813 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/765 "Passed tests") | [⏳ 🧪 api-mac](https://ews-build.webkit.org/#/builders/API-Tests-macOS-EWS "Waiting to run tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/1038 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/819 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/752 "Passed tests") | [⏳ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/GTK-WK2-Tests-EWS "Waiting to run tests") | 
| [❌ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/706 "Failed to checkout and rebase branch from PR 11084") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/727 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/775 "Passed tests") | [⏳ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/API-Tests-GTK-EWS "Waiting to run tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/793 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/764 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/737 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/832 "Built successfully") | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/4/builds/712 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/760 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/172 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/180 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/777 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/834 "Built successfully") | 
| | | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/166 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

darinadler reviewed (approved):
This change is fine and can land as is. Some tiny quibbles:

1. These two unrelated changes don’t need to be in the same pull request. It’s even better to do these one at a time. That’s a little funny coming from me since I often have huge pull requests, but it’s still best practice to do each change separate for a variety of reasons. Like if we have to revert something we revert as little as possible. Or if something has a surprising performance effect we know exactly which change it was.
2. The bug title is vague and the same could be used for many future pull requests that also change functions that only require `VM` rather than `JSGlobalObject`. It would be better to be slightly more specific in the title. This is another place where splitting in two is helpful. The FIXME part of the title is only accurate for half the pull request.
3. The bug title says "methods that only work with VMs" but that’s incorrect. More accurate would be something like "only require the VM, not the global object". The phrase "only work with VMs" means something else, like they would not work without a VM, which is true but not the point. It also doesn’t make much sense to say "the whole global object", because what we are passing is a pointer, not a whole object. The value here is that we don’t have to dereference a pointer, not that one object is smaller than the other.
4. The bug title and change log entry say "method", which is OK, but not quite right for C++. These are C++ **functions**. Some people use the word "method" for "member function", but I think it’s clearer to use C++ terminology instead.

LumioseSil committed:
Have functions that only require VMs take a VM reference https://bugs.webkit.org/show_bug.cgi?id=202392

Reviewed by Darin Adler.

We can have the .vm() method be called before these functions,
and pass the result VM reference instead of passing the pointer to the JSGlobalObject, potentially calling the .vm() methods redundantly in each method to which said JSGlobalObject pointer is passed, and dereferencing the pointers to the VM.

* Source/JavaScriptCore/runtime/JSONObject.cpp:(value): Change argument
  type from JSGlobalObject* to VM&.

* Source/WebCore/Modules/encryptedmedia/MediaKeyStatusMap.cpp:
  (get): Pass the VM of the state object to convertEnumerationToJS
  rather than the whole object.

* Source/WebCore/bindings/js/JSDOMConvertEnumeration.h:
  (convertEnumerationToJS): Change argument type from JSGlobalObject to
  VM.
  (convert): Pass the VM of lexicalGlobalObject instead of the entire
  global object.

* Source/WebCore/bindings/scripts/CodeGeneratorJS.pm: Ditto.

* Source/WebCore/bindings/scripts/test/JS/JSTestCallbackInterface.cpp:
  Ditto.

* Source/WebCore/bindings/scripts/test/JS/JSTestCallbackInterface.h:
  Ditto.

* Source/WebCore/bindings/scripts/test/JS/JSTestDefaultToJSONEnum.cpp:
  Ditto.

* Source/WebCore/bindings/scripts/test/JS/JSTestDefaultToJSONEnum.h:
  Ditto.

* Source/WebCore/bindings/scripts/test/JS/JSTestObj.cpp: Ditto.

* Source/WebCore/bindings/scripts/test/JS/JSTestObj.h: Ditto.

* Source/WebCore/bindings/scripts/test/JS/JSTestStandaloneDictionary.cpp:
  Ditto.

* Source/WebCore/bindings/scripts/test/JS/JSTestStandaloneDictionary.h:
  Ditto.

* Source/WebCore/bindings/scripts/test/JS/JSTestStandaloneEnumeration.cpp:
  Ditto.

* Source/WebCore/bindings/scripts/test/JS/JSTestStandaloneEnumeration.h:
  Ditto.

Canonical link: https://commits.webkit.org/262166@main
